src/deskolem.ml

```diff
@@ -3,186 +3,216 @@ open! Lplib
-    Invalid_argument otherwise. *)
-let get_option = fun opt ->
-    match opt with
-    | Some(x)   -> x
-    | None      -> raise (Invalid_argument "The option is None.")
+    | Meta _        -> assert false
+    | Patt _        -> assert false
+    | TEnv _        -> assert false
+    | Wild          -> assert false
+    | LLet _        -> assert false
+    | TRef _        -> assert false (* is not handled in the encoding. *)
 
 (** [size t] return the size of the term [t]. *)
 let rec size : term -> int = fun t ->
-  List.fold_right (fun a l -> size a + l) (snd (Term.get_args t)) 1
+  List.fold_right (fun a l -> size a + l) (Term.get_args t |> snd) 1
 
 (** [size_args f args] return the size of the term [f a0 a1 ... an] where args
     = [a0; a1; ...; an]. *)
 let size_args : sym -> term list -> int = fun f args ->
-    size (Term.add_args (mk_Symb(f)) args)
+    size (Term.add_args (mk_Symb f) args)
 
 (** [subst_var x m n] substitutes the variable [x] with the term [n] inside the
     term [m].*)
 let subst_var : term Bindlib.var -> term -> term -> term = fun x m n ->
-    Common.Console.out 4 "subst_var [ %a / %a] %a@." Print.pp_term n Print.pp_var x Print.pp_term m;
     Bindlib.subst (Bindlib.unbox (Bindlib.bind_var x (lift m))) n
 
-(** [subst_mvar x m n] substitutes all variables that [x] contains with the terms
-    of [n] inside the term [m]. *)
+(** [subst_mvar x m n] substitutes all variables that [x] contains with the 
+    terms of [n] inside the term [m]. *)
 let subst_mvar : term Bindlib.var list -> term -> term list -> term =
     fun x m n ->
     let x = Array.of_list x in
     let n = Array.of_list n in
     Bindlib.msubst (Bindlib.unbox (Bindlib.bind_mvar x (lift m))) n
 
-exception Not_Proof of term
-exception Not_ProofS of term
-
-(** [unProof t] returns the proposition which is inside the `Proof` constructor. *)
-let rec unProof : Sign.t -> term -> term = fun sign t ->
-    Common.Console.out 4 "unProof [%a]@." Print.pp_term t;
+(** [get_prop t] returns the proposition which is inside the `ϵ` constructor. *)
+let rec get_prop : Sign.t -> term -> term = fun sign t ->
     match unfold t with
-    |Appl(Symb({sym_name = "ϵ"; _}), t') -> t'
-    |Prod(a, b) ->
-        (try
-            let unProofA = try unProof sign a with Not_Proof _ -> raise (Not_ProofS a) in
-            let imp = Extra.StrMap.find "imp" !(sign.sign_builtins) in
-            assert (not (Bindlib.binder_occur b));
-            Term.add_args (mk_Symb imp) [unProofA; unProof sign (Bindlib.unbind b |> snd)]
-        with Not_ProofS _ ->
-            match a with
-            |Appl(Symb({sym_name = "τ"; _}), a') ->
-                let forall = Extra.StrMap.find "forall" !(sign.sign_builtins) in
-                Term.add_args (mk_Symb forall) [a'; unProof sign (Bindlib.unbind b |> snd)]
-            |_                                  ->
-                raise (Not_Proof t))
+    |Appl(Symb({sym_name = "ϵ"; _}), t')    -> t'
```

Comment:
This is not correct to test symbol name equality. You should use `Term.is_symb cfg.symb_proof` instead.
This seems also incorrect to take a signature as argument. This should be a config instead.